### DISCUSSION TITLE
UUIDs incorrectly stored

### DISCUSSION THREAD

**Discussion by ajdinmore** [directus/directus#6166] [Feature Requests]
UUIDs are designed to be stored as 128 bit values for index size/speed, but Directus is currently storing them strings which, in the case of MySQL, are indexed by only the first 4 bytes and can heavily impact performance.  I think the norm is to use e.g. `binary(16)`. 

I've only checked with MySQL, but I assume this is the case for all DBs without a native UUID type, if not across the board.

Sorry for being brief; I've not had a chance to look into the specifics of Directus' implementation yet and wanted raise the issue while I remembered.

**Comment by rijkvanzanten** [comment]
Directus relies on [`knex`](http://knexjs.org/#Schema-uuid) to set the type, which uses a native UUID type if it exists, or defaults to char(36) if it doesn't.

**Reply by ajdinmore** [reply]
Ah, thanks for the info.  I'll head over to their issues board with this then.

Kudos on Directus, I'm impressed with what I've seen so far.

**Reply by rijkvanzanten** [reply]
Thanks!

**Reply by grefrit** [reply]
@rijkvanzanten Is there any way to use binary(16) uuid with directus? Since we want to integrate it into an existing project and we cannot remake the database

**Reply by rijkvanzanten** [reply]
@GrefriT When using the "Generate UUID" option for primary keys in Directus, it will generate a UUIDv4 and tries inserting it as-is. IIRC a binary(16) representation of a UUID is a hexadecimal version of the UUID without hyphens right? Directus itself doesn't natively support that, but you should be able to use "Manually entered string", and then use a custom hook to generate that alternative UID format on creates. Everything else in the platform will work as it should 👍🏻

**Reply by grefrit** [reply]
@rijkvanzanten I understand that I can easily convert string to binary uuid when items.create.before fired, but I cannot convert binary uuid to string via items.read(items.read does not support .before) This is the whole problem. I already tried to use a hook for this, I managed to fix the uuid mappings in the collection, but I cannot modify the query.filter, and therefore if I open an item in the collection, then instead of the item there will be Page not found. It would be cool to have something like the ability to add a custom field transform

**Reply by rijkvanzanten** [reply]
> I understand that I can easily convert string to binary uuid when items.create.before fired, but I cannot convert binary uuid to string via items.read(items.read does not support .before) This is the whole problem.

Ah, well that wasn't obvious at all from your previous comment 😄 Sounds like the immediate solution would be to add .before modifier support to the read hook. Wanna open a  PR for that?

**Reply by grefrit** [reply]
@rijkvanzanten Yes, I will try to do it
